File: frontend/web/lib/stats.ts
export type Period = "day" | "week" | "month";

export interface SummaryStats {
  total_dialogs: number;
  active_users: number;
  avg_dialog_length: number;
}

export interface ActivityPoint {
  ts: string; // ISO string from API
  dialogs: number;
  messages: number;
}

export interface RecentDialog {
  dialog_id: string;
  user_id: string;
  started_at: string; // ISO string
  duration_sec: number;
  num_messages: number;
  status: string;
}

export interface TopUser {
  user_id: string;
  dialogs_count: number;
  messages_count: number;
  last_active_at: string; // ISO string
}

export interface DashboardStats {
  period: Period;
  summary: SummaryStats;
  activity: ActivityPoint[];
  recent_dialogs: RecentDialog[];
  top_users: TopUser[];
}

// Use internal URL for server-side requests, public URL for client-side
const getApiBaseUrl = (): string => {
  // Server-side: use internal Docker service name
  if (typeof window === 'undefined') {
    return process.env["STATS_API_URL_INTERNAL"] ?? "http://backend:8000/api/v1";
  }
  // Client-side: use public URL
  return process.env["NEXT_PUBLIC_STATS_API_URL"] ?? "http://localhost:8000/api/v1";
};

export const API_BASE_URL: string = getApiBaseUrl();

export async function getStats(period: Period): Promise<DashboardStats> {
  const url = `${getApiBaseUrl()}/stats?period=${period}`;
  const res = await fetch(url, { next: { revalidate: 60 } });
  if (!res.ok) {
    throw new Error(`Failed to load stats (${res.status})`);
  }
  const data = (await res.json()) as DashboardStats;
  return data;
}

export function formatDuration(seconds: number): string {
  const mins = Math.floor(seconds / 60);
  const secs = seconds % 60;
  return `${mins}m ${secs}s`;
}

File: src/api/routers/stats.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query

from api.collectors.mock import MockStatCollector
from api.models import DashboardStatsModel, Period


router = APIRouter(prefix="/api", tags=["stats"])


collector = MockStatCollector()


@router.get("/stats", response_model=DashboardStatsModel)
def get_stats(period: str = Query(..., description="day|week|month", examples={"day": {"value": "day"}})):
    normalized = period.lower()
    if normalized not in {p.value for p in Period}:  # case-insensitive enum check
        raise HTTPException(status_code=422, detail="Invalid period. Use day|week|month")

    stats = collector.get_stats(normalized)
    # Pydantic will validate and coerce the response model
    return {
        "period": stats.period,
        "summary": {
            "total_dialogs": stats.summary.total_dialogs,
            "active_users": stats.summary.active_users,
            "avg_dialog_length": stats.summary.avg_dialog_length,
        },
        "activity": [
            {"ts": p.ts, "dialogs": p.dialogs, "messages": p.messages} for p in stats.activity
        ],
        "recent_dialogs": [
            {
                "dialog_id": d.dialog_id,
                "user_id": d.user_id,
                "started_at": d.started_at,
                "duration_sec": d.duration_sec,
                "num_messages": d.num_messages,
                "status": d.status,
            }
            for d in stats.recent_dialogs
        ],
        "top_users": [
            {
                "user_id": u.user_id,
                "dialogs_count": u.dialogs_count,
                "messages_count": u.messages_count,
                "last_active_at": u.last_active_at,
            }
            for u in stats.top_users
        ],
    }


